hugr-core/src/extension/const_fold.rs

```diff
@@ -27,9 +142,29 @@ pub fn fold_out_row(consts: impl IntoIterator<Item = Value>) -> ConstFoldResult
 
 /// Trait implemented by extension operations that can perform constant folding.
 pub trait ConstFold: Send + Sync {
+    /// Given type arguments `type_args` and [`FoldVal`]s for each input,
+    /// update the outputs (these will be initialized to [FoldVal::Unknown]).
+    ///
+    /// Defaults to calling [Self::fold] with those arguments that can be converted ---
+    /// [FoldVal::LoadedFunction]s will be lost as these are not representable as [Value]s.
+    fn fold2(&self, type_args: &[TypeArg], inputs: &[FoldVal], outputs: &mut [FoldVal]) {
```

Comment:
why does this follow a mutable output pattern? Would some other process have set some values already that this one would leave unchanged?